cmd/util/cmd/find-trie-root/cmd.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot create WAL: %w", err)
+	}
+
+	from, to, err := prometheusWAL.Segments(dir)
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot get segments: %w", err)
+	}
+
+	if wantFrom > to {
+		return 0, 0, fmt.Errorf("from segment %d is greater than the last segment %d", wantFrom, to)
+	}
+
+	if wantTo < from {
+		return 0, 0, fmt.Errorf("to segment %d is less than the first segment %d", wantTo, from)
+	}
+
+	if wantFrom > from {
+		from = wantFrom
+	}
+
+	if wantTo < to {
+		to = wantTo
+	}
+
+	log.Info().
+		Str("dir", dir).
+		Int("from", from).
+		Int("to", to).
+		Int("want-from", wantFrom).
+		Int("want-to", wantTo).
+		Msgf("searching for trie root hash %v in segments [%d,%d]", expectedHash, wantFrom, wantTo)
+
+	sr, err := prometheusWAL.NewSegmentsRangeReader(lg, prometheusWAL.SegmentRange{
+		Dir:   w.Dir(),
+		First: from,
+		Last:  to,
+	})
+
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot create WAL segments reader: %w", err)
+	}
+
+	defer sr.Close()
+
+	reader := prometheusWAL.NewReader(sr)
+
+	for reader.Next() {
+		record := reader.Record()
+		operation, rootHash, _, err := wal.Decode(record)
+		if err != nil {
+			return 0, 0, fmt.Errorf("cannot decode LedgerWAL record: %w", err)
+		}
+
+		log.Debug().
+			Uint8("operation", uint8(operation)).
+			Str("root-hash", rootHash.String()).
+			Msgf("read LedgerWAL record")
+
+		switch operation {
+		case wal.WALUpdate:
+			if rootHash.Equals(expectedHash) {
+				log.Info().Msgf("found expected trie root hash %v", rootHash)
+				return reader.Segment(), reader.Offset(), nil
+			}
+		default:
+		}
+
+		err = reader.Err()
+		if err != nil {
+			return 0, 0, fmt.Errorf("cannot read LedgerWAL: %w", err)
+		}
+	}
+
+	return 0, 0, fmt.Errorf("finish reading all segment files from %d to %d, but not found", from, to)
+}
+
+func copyWAL(dir, outputDir string, segment int, expectedRoot ledger.RootHash) error {
+	writer, err := prometheusWAL.NewSize(log.Logger, nil, outputDir, wal.SegmentSize, false)
+	if err != nil {
+		return fmt.Errorf("cannot create writer WAL: %w", err)
+	}
+
+	defer writer.Close()
+
+	w, err := prometheusWAL.NewSize(log.Logger, nil, dir, wal.SegmentSize, false)
+	if err != nil {
+		return fmt.Errorf("cannot create WAL: %w", err)
+	}
+
+	sr, err := prometheusWAL.NewSegmentsRangeReader(log.Logger, prometheusWAL.SegmentRange{
+		Dir:   w.Dir(),
+		First: segment,
+		Last:  segment,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot create WAL segments reader: %w", err)
+	}
+
+	defer sr.Close()
+
+	reader := prometheusWAL.NewReader(sr)
+
+	for reader.Next() {
+		record := reader.Record()
+		operation, rootHash, update, err := wal.Decode(record)
+		if err != nil {
+			return fmt.Errorf("cannot decode LedgerWAL record: %w", err)
+		}
+
+		bytes := wal.EncodeUpdate(update)
+		_, err = writer.Log(bytes)
+		if err != nil {
+			return fmt.Errorf("cannot write LedgerWAL record: %w", err)
+		}
```

Comment:
`update` can be `nil` for delete operation record.  Maybe check operation before encoding record.